gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,447 @@
+      return this;
+    }
+
+    private KafkaWrapper build() {
+      Preconditions.checkArgument(!brokers.isEmpty(), "Need to specify at least one Kafka broker.");
+      return new KafkaWrapper(this);
+    }
+  }
+
+  private KafkaWrapper(Builder builder) {
+    this.useNewKafkaAPI = builder.useNewKafkaAPI;
+    this.brokers = builder.brokers;
+    this.kafkaAPI = getKafkaAPI();
+  }
+
+  public static KafkaWrapper create(State state) {
+    Preconditions.checkNotNull(state.getProp(KAFKA_BROKERS), "Need to specify at least one Kafka broker.");
+    return new KafkaWrapper.Builder().withAPIType(state.getPropAsBoolean(USE_NEW_KAFKA_API, DEFAULT_USE_NEW_KAFKA_API))
+        .withBrokers(state.getPropAsList(KAFKA_BROKERS)).build();
+  }
+
+  public List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist) {
+    return this.kafkaAPI.getFilteredTopics(blacklist, whitelist);
+  }
+
+  public long getEarliestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+    return this.kafkaAPI.getEarliestOffset(partition);
+  }
+
+  public long getLatestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+    return this.kafkaAPI.getLatestOffset(partition);
+  }
+
+  public Iterator<MessageAndOffset> fetchNextMessageBuffer(KafkaPartition partition, long nextOffset, long maxOffset) {
+    return this.kafkaAPI.fetchNextMessageBuffer(partition, nextOffset, maxOffset);
+  }
+
+  private KafkaAPI getKafkaAPI() {
+    if (this.useNewKafkaAPI) {
+      return new KafkaNewAPI();
+    } else {
+      return new KafkaOldAPI();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.kafkaAPI.close();
+  }
+
+  private abstract class KafkaAPI implements Closeable {
+    protected abstract List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist);
+
+    protected abstract long getEarliestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException;
+
+    protected abstract long getLatestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException;
+
+    protected abstract Iterator<MessageAndOffset> fetchNextMessageBuffer(KafkaPartition partition, long nextOffset,
+        long maxOffset);
+  }
+
+  /**
+   * Wrapper for the old low-level Scala-based Kafka API.
+   */
+  private class KafkaOldAPI extends KafkaAPI {
+    private static final int DEFAULT_KAFKA_TIMEOUT_VALUE = 30000;
+    private static final int DEFAULT_KAFKA_BUFFER_SIZE = 1024 * 1024;
+    private static final String DEFAULT_KAFKA_CLIENT_NAME = "kafka-old-api";
+    private static final int DEFAULT_KAFKA_FETCH_REQUEST_CORRELATION_ID = -1;
+    private static final int DEFAULT_KAFKA_FETCH_REQUEST_MIN_BYTES = 1024;
+    private static final int NUM_TRIES_FETCH_TOPIC = 3;
+    private static final int NUM_TRIES_FETCH_OFFSET = 3;
+
+    private final Map<String, SimpleConsumer> activeConsumers = Maps.newHashMap();
```

Comment:
Is is ever possible for an instance of this class to be used by multiple threads? If so, it's safer to use a `ConcurrentMap`.